app/scripts/modules/core/src/managed/constraints/constraintRegistry.tsx

```diff
@@ -0,0 +1,65 @@
+import React from 'react';
+import { DateTime } from 'luxon';
+
+import { relativeTime, timestamp } from '../../utils';
+import { IStatefulConstraint, StatefulConstraintStatus } from '../../domain';
+import { IconNames } from '../../presentation';
+
+const NO_FAILURE_MESSAGE = 'no details available';
+
+const throwUnhandledStatusError = (status: string) => {
+  throw new Error(`Unhandled constraint status "${status}", no constraint summary available`);
+};
+
+const { NOT_EVALUATED, PENDING, FAIL, OVERRIDE_PASS, OVERRIDE_FAIL } = StatefulConstraintStatus;
+
+interface IStatefulConstraintConfig {
+  iconName: IconNames;
+  shortSummary: (constraint: IStatefulConstraint) => JSX.Element | string;
+}
+
+export const isConstraintSupported = (type: string) => statefulConstraintOptionsByType.hasOwnProperty(type);
+export const getStatefulConstraintConfig = (type: string) => statefulConstraintOptionsByType[type];
+
+// Later, this will become a "proper" registry so we can separate configs
+// into their own files and extend them dynamically at runtime.
+// For now let's get more of the details settled and iterated on.
+export const statefulConstraintOptionsByType: { [type: string]: IStatefulConstraintConfig } = {
+  'manual-judgement': {
+    iconName: 'manualJudgement',
+    shortSummary: ({ status, judgedAt, judgedBy, startedAt, comment }: IStatefulConstraint) => {
+      const startedAtMillis = DateTime.fromISO(startedAt).toMillis();
+      const judgedAtMillis = DateTime.fromISO(judgedAt).toMillis();
+
+      switch (status) {
+        case NOT_EVALUATED:
+          return 'Manual judgement will be required before promotion';
+        case PENDING:
+          return (
+            <span>
+              Awaiting manual judgement since {relativeTime(startedAtMillis)}{' '}
+              <span className="text-italic text-regular sp-margin-xs-left">({timestamp(startedAtMillis)})</span>
+            </span>
+          );
+        case OVERRIDE_PASS:
+        case OVERRIDE_FAIL:
+          return (
+            <span className="sp-group-margin-xs-xaxis">
+              Manually {status === OVERRIDE_PASS ? 'approved' : 'rejected'} {relativeTime(judgedAtMillis)}{' '}
+              <span className="text-italic text-regular sp-margin-xs-left">({timestamp(judgedAtMillis)})</span>{' '}
+              <span className="text-regular">—</span> <span className="text-regular">by {judgedBy}</span>
+            </span>
+          );
+        case FAIL:
+          return (
+            <span className="sp-group-margin-xs-xaxis">
+              Something went wrong <span className="text-regular">—</span>{' '}
+              <span className="text-regular">{comment || NO_FAILURE_MESSAGE}</span>
+            </span>
+          );
+        default:
+          return throwUnhandledStatusError(status);
```

Comment:
Is there a way to let typescript do an exhaustive check of all the possible values of an enum instead of handling the `default` case?

Reply:
There is, but it has some disadvantages.

To my knowledge, the best way to get exhaustiveness here would be to have the handler return an object with functions for each constraint status. Then we could use a mapped type (`{ [status in StatefulConstraintStatus]: (constraint: IStatefulConstraint) => JSX.Element | string }`) to explicitly say "you have to return an object with a key for every value in this enum".

My hesitation in doing that is I'm not yet confident it makes sense to have a handler for every status, for every constraint. In this `manual-judgment` example, `PASS` is actually _not_ a state we expect to ever be possible because it always requires manual interaction (which produces an `OVERRIDE_PASS`/`OVERRIDE_FAIL` status). The backend code _can't_ actually send a manual-judgement into a `PASS` status. We wouldn't know how to handle it on the UI side as a result, and it doesn't seem like it makes much sense to put in UI code specifically for this constraint type intended to handle a case we don't think should exist or know how to represent. Instead I was thinking it'd be ideal for the core code to handle unexpected statuses in some common, generic way that might involve catching this error and showing that an unexpected case was hit with that constraint (i.e. treat it like a failure case, not a one-off thing that every constraint either did well or poorly).

Any opinions on whether that makes sense or not? There's certainly a counter-argument to be made for telling constraint developers they should handle every case and do some sort of best-effort attempt at presenting things in unexpected/impossible cases.

Reply:
I'm gonna think about doing the return-an-object thing in a subsequent PR, even if you have to return _every_ status in that object (i.e. the type can be partial)